osd,rados: add repair-copy op for scrub/repair

tchaikov opened the pull request:
this PR is based on #8931 by @dzafman

tchaikov commented:
tests on their way.

tchaikov commented:
per sam

> i think it would be simpler to just allow the repair_write call to specify a set of bad shards. For replicated pool, we simply choose one which is not in that set . For EC pools, we use that information to avoid bad shards.

tchaikov commented:
@dzafman  and @athanatos could you take a look ? next, i will add the test for ec pool `repair-copy` .

tchaikov commented on src/osd/ReplicatedPG.cc:
```diff
@@ -1357,6 +1357,35 @@ int ReplicatedPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
+  uint32_t what;
+  vector<pg_shard_t> bad_shards;
+  try {
+    ::decode(what, bp);
```
maybe i should wrap it in an arg struct, and add the `encode()/decode()` methods for it.

athanatos replied:
Yeah, probably.

tchaikov commented on src/tools/rados/rados.cc:
```diff
@@ -1397,7 +1398,10 @@ static void dump_inconsistent(const inconsistent_obj_t& inc,
-    f.dump_int("osd", osd_shard.first);
+    f.dump_int("osd", osd_shard.first.first);
+    auto shard = osd_shard.first.second;
+    if (shard)
```
should be `if (shard == -1)`

athanatos commented on src/osd/ReplicatedPG.cc:
```diff
@@ -7075,7 +7075,17 @@ int ReplicatedPG::fill_in_copy_get(
       cursor.data_complete = true;
       dout(20) << " got data" << dendl;
     }
-    assert(cursor.data_offset <= oi.size);
```
I don't follow how this can happen.  Isn't data_offset always 0 or a result of a previous copy_get?

tchaikov replied:
see https://github.com/ceph/ceph/pull/9203/files#diff-72747d40a424e7b5404366b557ff12a3L7053 . it could be greater than expected, if we are reading a bad replica which is larger than what object-info claims.

athanatos replied:
I see, gotcha.

athanatos commented on src/include/rados/librados.hpp:
```diff
@@ -712,6 +712,14 @@ namespace librados
+     * @param version the version of rewritten object
+     * @param what the flags indicating what we will copy
+     */
+    int repair_copy(const std::string& oid, uint64_t version, uint32_t what, int32_t osd, uint32_t epoch);
```
This bit should probably be squashed into the later commit which changes it to take a set of bad shards, right?

athanatos commented on src/include/rados/librados.hpp:
```diff
@@ -715,11 +715,17 @@ namespace librados
      */
-    int repair_copy(const std::string& oid, uint64_t version, uint32_t what, int32_t osd, uint32_t epoch);
+    int repair_copy(const std::string& oid, uint64_t version, uint32_t what,
+                    const std::vector<std::pair<int32_t, int8_t>>& bad_osds,
```
Do we need to expose the <osd, shard> tuple anywhere else?  Probably worth exposing a librados type to wrap it.

tchaikov commented on src/osd/ReplicatedPG.h:
```diff
@@ -1346,14 +1355,18 @@ class ReplicatedPG : public PG, public PGBackend::Listener {
-   * @param src: The source object
    * @param oloc: the source object locator
    * @param version: the version of the source object to copy (0 for any)
+   * @param what: the flags composed of CopyOp::COPY_* indicating what to copy
```
revert this change.

tchaikov commented on src/osd/ReplicatedPG.h:
```diff
@@ -1346,14 +1355,18 @@ class ReplicatedPG : public PG, public PGBackend::Listener {
    *
    * @param cb: The CopyCallback to be activated when the copy is complete
    * @param obc: The ObjectContext we are copying into
-   * @param src: The source object
```
revert this change.

tchaikov commented on src/osd/ReplicatedPG.h:
```diff
@@ -1346,14 +1355,18 @@ class ReplicatedPG : public PG, public PGBackend::Listener {
+  void start_copy(CopyCallback *cb, ObjectContextRef obc,
+		  hobject_t src,
+		  object_locator_t oloc, version_t version,
+		  unsigned flags,
```
and this.

tchaikov commented:
@dzafman i introduced a new librados type `osd_shard_t ` per the suggest of @athanatos, and updated the commits for adding user_version and shard_id. maybe you want to cherry-pick them again?

tchaikov commented:
i am still testing the test for ec pool.

tchaikov commented:
@athanatos @dzafman comments addressed. because these two tests are time consuming, i will move them into qa units. so we can exercise them in the integration tests. what do you think?

athanatos commented:
I would be fine with moving them into qa units.

dachary commented:
needs rebase

dachary commented:
setting DNM since the base PR is not merged yet

tchaikov commented:
@dzafman and @athanatos could you take a look again? once it looks good, i will move replace `faster_objectstore_tool()` with `objectstore_tool()` and move the newly added tests into "qa".

changelog

- @athanatos 's comments at https://github.com/ceph/ceph/pull/8931#issuecomment-219499271 are addressed
- rebased against master
- updated the commits from #8931 accordingly

tchaikov commented:
```
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:1151: wait_for_clean:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:137: rados_put_get:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:172: rados_get_data_eio:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/erasure-code/test-erasure-eio.sh:313: TEST_rados_get_with_subreadall_eio_shard_0:  return 1
```

tchaikov commented:
@dzafman i tested the failed test case locally and it passed.

tchaikov commented:
retest this please.

tchaikov commented:
rebased against master.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -1529,6 +1529,46 @@ int PrimaryLogPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
+  const auto& oi = ctx->obs->oi;
+  if (!ctx->copy_cb) {
+    // mark the bad shard missing so they are not chosen when repairing
+    if (pool.info.require_rollback()) {
```
I think here you really mean pool.info.ec_pool().  Does it actually matter though?  Why not add it to the peer_missing sets unconditionally so that Replicated and EC work uniformly?

tchaikov commented on src/librados/IoCtxImpl.cc:
```diff
@@ -1247,13 +1287,14 @@ int librados::IoCtxImpl::get_inconsistent_objects(const pg_t& pg,
     nullptr, nullptr);
   objecter->op_submit(o, &c->tid);
-  return 0;
+  return rval;
```
fixed.

tchaikov commented on src/librados/IoCtxImpl.cc:
```diff
@@ -1247,13 +1287,14 @@ int librados::IoCtxImpl::get_inconsistent_objects(const pg_t& pg,
   ::ObjectOperation op;
-  op.scrub_ls(start_after, max_to_get, objects, interval, nullptr);
+  int rval = 0;
+  op.scrub_ls(start_after, max_to_get, objects, interval, &rval);
```
right.

tchaikov commented on src/librados/IoCtxImpl.cc:
```diff
@@ -780,6 +780,46 @@ int librados::IoCtxImpl::operate_read(const object_t& oid,
+  Objecter::Op *objecter_op = objecter->prepare_read_op(oid, oloc,
+	                                      *o, snap_seq, NULL,
+	                                      flags | CEPH_OSD_FLAG_REPAIR_READS | CEPH_OSD_FLAG_IGNORE_OVERLAY,
+	                                      onack, &ver);
```
fixed.

tchaikov commented:
@athanatos all comments addressed and repushed.

athanatos commented on src/librados/IoCtxImpl.cc:
```diff
@@ -780,6 +780,44 @@ int librados::IoCtxImpl::operate_read(const object_t& oid,
+  objecter_op->target.use_osd_epoch = true;
+  objecter_op->target.epoch = e;
+  objecter->op_submit(objecter_op, &c->tid);
+  return c->tid;
```
I'm fairly sure the expectation is that the return value for these aio_operate* calls should be 0 if successfully initiated.  Getting the tid back I think would be unexpected.

athanatos commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -1529,6 +1529,44 @@ int PrimaryLogPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
+    return -EINPROGRESS;
+  } else {
+    // finish
+    if (pool.info.require_rollback()) {
```
You unconditionally added them to the missing sets, right?  Seems like you need to unconditionally remove them?

tchaikov commented:
@athanatos all comments addressed and repushed.

dzafman commented on src/osdc/Objecter.h:
```diff
@@ -2328,6 +2340,13 @@ class Objecter : public md_config_obs_t, public Dispatcher {
+  /**
+   * get a random osd not listed in blacklist
+   * @note the primary osd will be returned even it is listed in the blacklist
+   *       if the target pool is a erasure pool, the primary osd is returned.
```
Confusing because you mention primary osd twice.  Is this sufficient?

the primary osd will be returned even it is listed in the blacklist
if the target pool is a erasure pool

dzafman commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -5946,6 +5987,24 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 
+    case CEPH_OSD_OP_REPAIR_COPY:
+      ++ctx->num_write;
+      result = do_repair_copy(ctx, osd_op, bp);
```
No tracepoint here?

dzafman commented on src/tracing/osd.tp:
```diff
@@ -790,3 +790,15 @@ TRACEPOINT_EVENT(osd, do_osd_op_post,
     )
 )
+
+TRACEPOINT_EVENT(osd, do_osd_op_pre_assert_interval,
```
If you add tracepoint to repair copy don't forget to add here too.

dzafman commented on src/tools/rados/rados.cc:
```diff
@@ -322,6 +328,94 @@ static int do_get(IoCtx& io_ctx, RadosStriper& striper,
+  }
+  epoch_t epoch = atoi(epoch_str);
+  uint64_t version = atoi(version_str);
+  uint32_t what = (librados::repair_copy_t::DATA |
```
Shouldn't we include "omap" in the default or just "data" as the default?  I feel like the command line format is limiting because you can only specify 1 of the elements.

tchaikov commented on src/librados/librados.cc:
```diff
@@ -1930,6 +1965,20 @@ int librados::IoCtx::aio_sparse_read(const std::string& oid, librados::AioComple
 
+int librados::IoCtx::aio_repair_read(const std::string& oid, librados::AioCompletion *c,
+				     bufferlist *pbl, size_t len, uint64_t off,
+				     uint64_t /*snapid*/, int flags, int32_t osdid, epoch_t e)
```
can we track this in the tracker and support it in another PR?

jdurgin replied:
is snap_seq passed in by aio_operate_repair_read() not respected?

tchaikov commented:
@dzafman comments addressed and repushed.

dzafman commented:
@tchaikov Everything looks good, but the tracepoint change isn't building

tchaikov commented:
changelog

- squashed related commits,
- rebased against master

tchaikov commented:
being tested at http://pulpito.ceph.com/kchai-2017-01-30_09:44:58-rados-wip-kefu-testing---basic-mira/

tchaikov commented:
@athanatos the tests passed, is this change good to merge?

tchaikov commented:
hey @jdurgin could you please take a look?

jdurgin commented:
certainly, I started earlier this week and will finish looking through it today

jdurgin commented on src/librados/librados.cc:
```diff
@@ -1266,6 +1266,26 @@ int librados::IoCtx::read(const std::string& oid, bufferlist& bl, size_t len, ui
+				 size_t len, uint64_t off, int flags,
+				 int32_t osdid, epoch_t e)
+{
+  auto c = librados::Rados::aio_create_completion();
```
this could be simplified by calling operate_repair_read()

tchaikov replied:
```diff
@@ -1268,19 +1268,16 @@ int librados::IoCtx::read(const std::string& oid, bufferlist& bl, size_t len, ui

 int librados::IoCtx::repair_read(const std::string& oid, bufferlist& bl,
                                 size_t len, uint64_t off, int flags,
-                                int32_t osdid, epoch_t e)
+                                int32_t osdid, epoch_t epoch)
 {
-  auto c = librados::Rados::aio_create_completion();
-  int r = aio_repair_read(oid, c, &bl, len, off, 0, flags, osdid, e);
-  if (r < 0) {
-    c->release();
-    return r;
-  }
-  c->wait_for_complete();
-  r = c->get_return_value();
-  if (r >= 0)
-    io_ctx_impl->set_sync_op_version(c->get_version64());
-  c->release();
+  if (len > (size_t) INT_MAX)
+    return -EDOM;
+  ObjectReadOperation op;
+  io_ctx_impl->prepare_assert_ops(&op.impl->o);
+  op.read(off, len, &bl, nullptr);
+  // Only 1 flag can be passed from user
+  flags &= CEPH_OSD_OP_FLAG_FAILOK;
+  int r = operate_repair_read(oid, &op, &bl, osdid, epoch, flags);
   if (r < 0)
     return r;
   return bl.length();
```

i'd say it's more of a trade-off. probably, the call path looks more reasonable this way: 

before this change:
- repair_read() => aio_repair_read()
- aio_repair_read() => aio_operate_repair_read()

after this change:
- repair_read() => operate_repair_read() => aio_operate_repair_read(
- aio_repair_read() => aio_operate_repair_read()

jdurgin replied:
I agree, the first way was duplicating fewer unique things.

In general these librados.cc methods should be doing as little as possible - just the necessities to translate the C++ and C interfaces to a common internal call in IoCtxImpl. You'll notice nothing else in librados.cc calls prepare_assert_ops() for example, all that sort of common code is in IoCtxImpl. This bit can be cleaned up when we add the C interface though, this pr doesn't have to do everything at once.

jdurgin commented on src/osd/PG.cc:
```diff
@@ -5270,7 +5270,9 @@ bool PG::can_discard_op(OpRequestRef& op)
   }
 
-  if (m->get_map_epoch() < pool.info.last_force_op_resend &&
+  // Don't care about last_force_op_resend for recovery reads
```
why not? seems like a bug

tchaikov replied:
@dzafman neither do i understand this. could you elaborate a little bit?

tchaikov commented on src/include/rados/librados.hpp:
```diff
@@ -1058,6 +1078,8 @@ namespace librados
     int operate(const std::string& oid, ObjectWriteOperation *op);
     int operate(const std::string& oid, ObjectReadOperation *op, bufferlist *pbl);
+    int operate_repair_read(const std::string& oid, librados::ObjectReadOperation *o, bufferlist *pbl,
+		       uint32_t osdid, int32_t epoch, int op_flags);
```
`op_flags` is different from `flags` in this context, `flags` is the bitset of `librados::OPERATION_*`, and it's a part of our public API. while `op_flags` is the internal representation of it, which is part of the rados protocol, and is understood by OSD. we use `translate_flags()` to translate from the former to the later. so i'd prefer to keep `op_flags` , so it's more explicit, and less misleading. 

what do you think?

tchaikov commented on src/librados/IoCtxImpl.cc:
```diff
@@ -1389,6 +1427,44 @@ int librados::IoCtxImpl::read(const object_t& oid,
+  comp->io = this;
+  queue_aio_write(comp);
+
+  auto objecter_op = objecter->prepare_mutate_op(oid, oloc, op, snapc, ut, 0,
```
fixed.

tchaikov commented:
apart from https://github.com/ceph/ceph/pull/9203#discussion_r99285234, all other comments addressed.

tchaikov commented on src/librados/IoCtxImpl.cc:
```diff
@@ -799,16 +797,12 @@ int librados::IoCtxImpl::aio_operate_repair_read(const object_t& oid,
-  o->out_rval.resize(size);
-  o->out_bl[1] = pbl;
-  o->ops[1].op.flags = op_flags;
+  o->assert_interval(e);
```
@dzafman i changed the way how you prepend the op, because i think this is less hackery.

jdurgin commented on src/librados/IoCtxImpl.cc:
```diff
@@ -780,6 +780,40 @@ int librados::IoCtxImpl::operate_read(const object_t& oid,
+
+  int op = o->ops[0].op.op;
+  ldout(client->cct, 10) << ceph_osd_op_name(op) << " oid=" << oid << " nspace=" << oloc.nspace << dendl;
+  // Prepend assert_interval op
```
instead of modifying the ObjectOperation, you can use the extra_ops arg for the Objecter prepare_* helpers.
adding an Objecter::prepare_repair_read() helper here would be good too

tchaikov replied:
> instead of modifying the ObjectOperation, you can use the extra_ops arg for the Objecter prepare_* helpers.

it's a different use case. `prepare_read_op()` prepend the existing ops in `extra_ops` to the new read op allocated in it. but we want to prepend an `assert_interval` op to the existing read ops in `o`.

> adding an Objecter::prepare_repair_read() helper here would be good too

okay.

jdurgin commented on src/osdc/Objecter.cc:
```diff
@@ -2736,6 +2774,9 @@ int Objecter::_calc_target(op_target_t *t, bool any_change)
 	t->sort_bitwise,
 	sort_bitwise,
 	pg_t(prev_seed, pgid.pool(), pgid.preferred()))) {
+    // If the interval has changed, then don't resend for repair read
```
won't this make the librados op hang? seems like we should resend, and let the osd give us the error if the interval changed

tchaikov replied:
right! and it's simpler.

jdurgin commented on src/osdc/Objecter.h:
```diff
@@ -1206,10 +1216,12 @@ class Objecter : public md_config_obs_t, public Dispatcher {
     int osd = -1;      ///< the final target osd, or -1
 
     epoch_t last_force_resend = 0;
+    epoch_t epoch; ///< Required epoch
```
how about naming it required_interval (since we care about intervals changing, not a new map that doesn't affect our pg), defaulting to 0, and getting rid of use_osd_epoch?

tchaikov replied:
maybe we can just leave the interval checking to `assert_interval` op, and instead of using `required_interval` or `use_osd_epoch`, we can have a `force_use_osd` to specify the designated osd so object will always use it as long as it's up.

tchaikov commented on src/osdc/Objecter.cc:
```diff
@@ -2642,6 +2642,44 @@ int64_t Objecter::get_object_pg_hash_position(int64_t pool, const string& key,
+    if (acting.empty()) {
+      return -1;
+    }
+    int p = rand() % acting.size();
```
okay. but i think it would fine if we always use the same sequence of the pseudo random number in every Objecter. but i don't feel strong either way. will use `random()` and `srandom()` instead for better randomness.

> POSIX.1-2008 marks rand_r() as obsolete.

quote from rand(3).

jdurgin replied:
sounds good. it doesn't matter a ton for this use, but better to be safe for future uses. as long as it's not the ridiculous c++ api http://www.pcg-random.org/posts/cpps-random_device.html

jdurgin commented on src/test/osd/osd-scrub-repair.sh:
```diff
@@ -87,6 +92,202 @@ function TEST_corrupt_and_repair_replicated() {
+#
+# Corrupt one copy of a replicated pool and repair it manually
+#
+# TODO: move this test to ceph-qa-suite
```
we need at least a ceph_test_librados_api* test to get coverage with the various qa-suite options (esp. failure injection for the objecter resending stuff)

ideally there would be RadosModel tests for the new ops as well

tchaikov replied:
if this test is moved to `qa/suites/rados/basic/tasks`, i think it will have the coverage of the various qa-suite options.

maybe we can postpone this to another PR? i'd really want to have this feature in Luminous if possible.

tchaikov commented on src/include/rados/librados.hpp:
```diff
@@ -1058,6 +1078,8 @@ namespace librados
     int operate(const std::string& oid, ObjectReadOperation *op, bufferlist *pbl);
+    int operate_repair_read(const std::string& oid, librados::ObjectReadOperation *o,
+		       uint32_t osdid, int32_t epoch);
     int aio_operate(const std::string& oid, AioCompletion *c, ObjectWriteOperation *op);
```
will expose `aio_operate_repair_read()` instead. user is always able to build up his/her own `operate_repair_read()` this method.

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -1529,6 +1529,47 @@ int PrimaryLogPG::do_scrub_ls(MOSDOp *m, OSDOp *osd_op)
+  if (!ctx->copy_cb) {
+    // mark the bad shard missing so they are not chosen when repairing
+    for (const auto& bad_shard : arg.bad_shards) {
+      peer_missing[bad_shard].add(oi.soid, oi.version, eversion_t());
```
erasure pool select source shards from (acting_set -  peer_missing).

i don't follow, maybe the question should be "what makes it unsafe to use peer_missing in this way?".

jdurgin replied:
peer_missing is used in many other places too (e.g. lost_unfound_revert, backfill, recovery) - I'm worried one of those would be affected by adding temporary entries to peer_missing. if the only reason it's used here is to exclude certain shards, how hard is it to pass that information down explicitly?

tchaikov replied:
i did investigate the possible ways to pass this down. it's possible, but complicates the code path a lot. marking the corrupted shard missing fits our needs pretty well.

> I'm worried one of those would be affected by adding temporary entries to peer_missing.

i will investigate this. and come back to you.

tchaikov commented on src/include/rados/librados.hpp:
```diff
@@ -757,6 +757,21 @@ namespace librados
     int writesame(const std::string& oid, bufferlist& bl,
 		  size_t write_len, uint64_t off);
     int read(const std::string& oid, bufferlist& bl, size_t len, uint64_t off);
+    int repair_read(const std::string& oid, bufferlist& bl, size_t len, uint64_t off, int flags, int32_t osdid, uint32_t e);
```
i will add the C and python bindings once the C++ version is stablized.

tchaikov commented on src/librados/IoCtxImpl.cc:
```diff
@@ -780,6 +780,40 @@ int librados::IoCtxImpl::operate_read(const object_t& oid,
+				      int flags, int32_t osdid, epoch_t e)
+{
+  if (!o->size())
+    return 0;
```
removed.

tchaikov commented on src/include/rados/librados.hpp:
```diff
@@ -757,6 +757,21 @@ namespace librados
     int writesame(const std::string& oid, bufferlist& bl,
 		  size_t write_len, uint64_t off);
     int read(const std::string& oid, bufferlist& bl, size_t len, uint64_t off);
+    int repair_read(const std::string& oid, bufferlist& bl, size_t len, uint64_t off, int flags, int32_t osdid, uint32_t epoch);
```
not really, the result will be stored to the operation's own rval. so it's still good to use it?

jdurgin replied:
the overall return value (all that's inspected in repair_copy() or the non-operation versions of repair_read) will be 0 even if the interval assert on the repair_read fails, right?

tchaikov replied:
right.

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -1866,7 +1908,8 @@ void PrimaryLogPG::do_op(OpRequestRef& op)
   // missing object?
-  if (is_unreadable_object(head)) {
+  if (is_unreadable_object(head) &&
+      !(m->get_flags() & CEPH_OSD_FLAG_REPAIR_READS)) {
```
done.

tchaikov commented on src/osdc/Objecter.cc:
```diff
@@ -2642,6 +2642,44 @@ int64_t Objecter::get_object_pg_hash_position(int64_t pool, const string& key,
+int32_t Objecter::pick_random_osd(const op_target_t& t,
+				  const vector<pg_shard_t>& blacklist) const
+{
+  const pg_pool_t *pi = osdmap->get_pg_pool(t.base_oloc.pool);
```
done.

tchaikov commented on src/osd/PG.h:
```diff
@@ -692,7 +692,8 @@ class PG : protected DoutPrefixProvider {
   map<pg_shard_t, pg_info_t>    peer_info;   // info from peers (stray or prior)
   set<pg_shard_t> peer_purged; // peers purged
-  map<pg_shard_t, pg_missing_t> peer_missing;
+  map<pg_shard_t, pg_missing_t> peer_missing; // peer_missing[primary] is used
```
fixed.

tchaikov commented on src/librados/IoCtxImpl.cc:
```diff
@@ -1389,6 +1423,45 @@ int librados::IoCtxImpl::read(const object_t& oid,
+  comp->io = this;
+  queue_aio_write(comp);
+
+  auto objecter_op = objecter->prepare_mutate_op(oid, oloc, op, snapc,
```
`Objecter::init_ops()` prepend the ops in given `ObjectOperation` into the new ops.  like
```
[extra,extra,extra,to-be-filled]
```
`prepare_read_op()` fills up the last "to-be-filled slot with a newly created read op.

but in the repair-read case, what we want is
```
[assert-interval,extra,extra,extra]
```
where extra is the read ops passed in by caller. to use `init_ops()` we need to create an `ObjectOperation`, and put assert-interval into it first, and copy the ops in `ObjectOperation` to the ops filled by `init_ops()`. i don't see the advantage of this approach.

jdurgin replied:
keeping it as is sounds good, can clean up later if an opportunity appears

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -7110,6 +7185,45 @@ void PrimaryLogPG::start_copy(CopyCallback *cb, ObjectContextRef obc,
+  auto cb = new CopyFromCallback(ctx);
+  ctx->copy_cb = cb;
+  uint32_t to_copy = 0;
+  if (what & librados::repair_copy_t::DATA)
```
i guess it's fine at this moment: the internal representation can share with the same definitions defined in public interface. and i think it's desirable. if they diverge over the time, we certainly have the luxury to use another set of protocol internally and do the translation in librados. and we are also using other stuff in the `librados` in `PrimaryLogPG.cc` if you search `librados` in that source file.

jdurgin replied:
I'd consider those other uses bugs - the fadvise flags in fact have CEPH_OSD_OP_FLAG...* constants defined, and that's important since that header is shared with the kernel.

The other current user is the 2nd librados listing interface, which imo is another misuse of the librados namespace, since it's used in the wire protocol for pgnls. 

We shouldn't be able to break the wire protocol accidentally by changing a librados type.

tchaikov replied:
okay. will fix this.

tchaikov replied:
done.

tchaikov commented on doc/rados/troubleshooting/troubleshooting-pg.rst:
```diff
@@ -452,9 +463,28 @@ some predefined criteria. But this does not always work. For example, the stored
+from osd.0 at epoch 14. If you are happy with this replica after looking at
+it, you might want to overwrite bad replicas with good ones::
+
+  rados -p rbd repair-copy foo 2 14 1
```
if the interval changes or it is overwritten with new bits, how can we know where the object is not located at, or if its replicas/shards are consistent or not without a scrubbing + get-inconsistent call?

jdurgin replied:
consider an admin who detects a bad memory module, replaces it, and then scrubs to find that there are inconsistencies on the formerly bad node.

If there are random writes happening to the inconsistent objects, and no other cluster changes, the admin knows the copies on the formerly bad node still need repairing. Having to rescrub after each write seems pretty inconvenient, and may never allow repair this way if there are more random writes before the admin can use repair-copy with the new scrub output.

This sort of situation makes me wonder if repair-copy should avoid the version assert entirely - what benefit does it provide beyond the interval assert?

tchaikov replied:
the version assert to prevent us from overwriting the changed object with the old copy.

jdurgin replied:
what about having the repair-copy op pull the latest version, like copy-from does?

This assumes the caller (eventually ceph-mgr) is repairing the object's metadata (object info, snapset, clones) before using this call to repair its data/omap/xattrs - and in that case we can trust the local object info's version

tchaikov replied:
@jdurgin i don't think ceph-mgr will be (eventually) smart enough to "repair" the consistencies. instead i think the expected caller of these API is user's own script. it (the script) will be able to inspect the content of object, to check the integrity of it, and to choose the correct one using user's domain knowledge about the content **before** using repair-copy to overwrite the corrupted ones with golden copy.

tchaikov commented:
@jdurgin are you good with the fixes in the `[to-be-squashed]` commits? if yes, i will squash them into the corresponding commits. so i can have less overhead maintaining the commits.

tchaikov commented:
i will make sure all comments are addressed after squashing these "to-be-squashed" commits.

jdurgin commented on src/osdc/Objecter.cc:
```diff
@@ -2812,9 +2809,9 @@ int Objecter::_calc_target(op_target_t *t, bool any_change)
+      if (t->forced_osd != -1) {
 	assert(read);
-	osd = osdmap->is_up(t->osd) ? t->osd : -1;
+	osd = osdmap->is_up(t->forced_osd) ? t->forced_osd : -1;
```
hmm, even if the osd is up, it may no longer hold the pg for the repair read op - in which case it'll drop the op.

I suppose we could check if it's in acting here, and if not cancel the repair read request with ERANGE on the client side... a bit ugly to add that to the client though...

tchaikov replied:
thanks and agreed! will fix.

tchaikov replied:
done.

jdurgin commented:
@tchaikov squashing the existing commits is fine with me

jdurgin commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -6006,13 +6006,10 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
 	tracepoint(osd, do_osd_op_pre_assert_interval, soid.oid.name.c_str(), soid.snap.val, epoch);
 	dout(5) << "Check interval " << soid << " epoch " << epoch << " pg epoch " << pg_epoch << dendl;
-	if (epoch > pg_epoch) {
+	if (epoch != pg_epoch) {
```
this one is confusing to me - won't the epoch be different if there have been any osdmap changes that don't affect this pg? or is scrub reporting the same_interval_since epoch?

tchaikov replied:
yes, scrub is reporting the same_interval_since epoch.

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -7165,12 +7283,32 @@ void PrimaryLogPG::_copy_some(ObjectContextRef obc, CopyOpRef cop)
+    // we are holding the write lock of the object being repaired, so, to read from
+    // it, we need to SKIPRWLOCKS here. and since we are the very one holding the
+    // write lock, it's safe to skip the rwlock.
+    auto objecter_op = osd->objecter->prepare_read_op(cop->src.oid, cop->oloc, op,
```
fixed.

tchaikov commented:
@jdurgin all comments addressed and repushed.

jdurgin commented on src/librados/librados.cc:
```diff
@@ -2244,7 +2259,9 @@ int librados::IoCtx::repair_copy(const string& oid, uint64_t version, uint32_t w
   }
-  return io_ctx_impl->repair_copy(object_t(oid), version, what, shards, epoch);
+  return io_ctx_impl->repair_copy(object_t(oid), version,
+				  translate_repair_copy(what),
```
looks good, just need to add to the other versions of repair_copy

tchaikov commented:
see http://pad.ceph.com/p/scrub-repair, closing for now.
